### PR TITLE
[REVIEW] Parquet reader: add support for multiple pandas index columns

### PR DESCRIPTION
**Pull request by OlivierNV**
When loading a subset of columns and adding the pandas index, the parquet reader currently assumes a single index column.

### CODE REVIEW

File: cpp/src/io/parquet/reader_impl.cu [hunk 1]
```diff
@@ -220,24 +220,41 @@ struct metadata : public FileMetaData {
...
 
     if (it != key_value_metadata.end()) {
-      const auto pos = it->value.find("index_columns");
-      if (pos != std::string::npos) {
-        const auto begin = it->value.find('[', pos);
-        const auto end = it->value.find(']', begin);
-        if ((end - begin) > 1) {
-          return it->value.substr(begin + 2, end - begin - 3);
+      const auto index_start = it->value.find("index_columns");
+      if (index_start != std::string::npos) {
```
[the remark this answers is not in the document]
Comment (OlivierNV) [reply cpp/src/io/parquet/reader_impl.cu]: Okay, will look into that...

**Comment by OlivierNV** [comment]
Unfortunately, it looks like the regex version is causing the parquet test to fail. Need to debug this.

**Comment by OlivierNV** [comment]
I'm going to have to revert the use of regex for parsing index column names. From what I can tell, it just won't work, as there could be other sections that include brackets in the string, and there could be non-word characters in column names (as well as newlines randomly distributed within the JSON string).

**Comment by harrism** [comment]
It can definitely work...

**Comment by OlivierNV** [comment]
> It can definitely work...

I'm all ears: I tried over at regex101, but just to extract the "pandas_index": [(*)], every solution I found that worked for one case would easily break for other cases.

**Comment by harrism** [comment]
So in the JSON (?) you posted, it's `"index_columns":...` instead of `index_columns = ...` as in the code. Which one is it that you need to parse?

In any case, if you know the list of column names must end in a double quote and a square bracket, you can use this regex which stops at the first closing bracket after a quote...

Tested with \n, whitespace, brackets, and lots of other stuff inside the quotes...
https://regex101.com/r/J3sjin/3

(.*\"index_columns\"\s*=\s*\[\s*(\".*\")\s*\])

**Comment by harrism** [comment]
BTW if you don't want quotes around `index_columns`, this is the regex: `(.*index_columns\s*=\s*\[\s*(\".*\")\s*\])`

**Comment by OlivierNV** [comment]
> BTW if you don't want quotes around `index_columns`, this is the regex: `(.*index_columns\s*=\s*\[\s*(\".*\")\s*\])`

Quotes are not the problem. The main problem is that there are subsequent sections in the JSON string that have square brackets, and regex seems to capture everything between the first and last square bracket in the entire string (which basically returns the entire JSON string, it can be mitigated by requiring a separator following the ], but that breaks if the index is the last entry in the string, and also did not handle an empty index like [], so the whole thing looks extremely fragile from what I can tell. I couldn't find something to tell regex to capture everything within the first two square brackets in the input string.

**Comment by harrism** [comment]
I think you missed the point of my change. It adds explicit quote matching inside the square brackets, which forces it to stop at the first closing square bracket after a quote. Since you can't have non-escaped quotes inside a quoted string, this seems robust.

**Comment by harrism** [comment]
If the index_columns part can appear anywhere in the initial metadata string, then this is the regex you want: https://regex101.com/r/J3sjin/4

```
(index_columns\s*=\s*\[\s*(\".*\")\s*\])
```

**Comment by OlivierNV** [comment]
> ```
> (index_columns\s*=\s*\[\s*(\".*\")\s*\])
> ```

Does not work for empty square brackets, eg: "index_columns": []
(It's not an equal btw, it's a colon, my initial example was wrong). I'm not sure how pandas handles quotes or brackets within column names ~ mainly concerned with basic cases at this point.

**Comment by harrism** [comment]
Working on it. I'm sure there's a regex for this. :)

**Comment by OlivierNV** [comment]
With a bit of googling around, found a potential candidate that seems to handle all 3 cases, though it includes the square brackets in the returned string:
\"index_columns\"[ :]+((?=\[)\[[^]]*\]|(?=\{)\{[^\}]*\}|\"[^\"]*\")
Willing to give it a shot, but good luck fixing a bug in that, though :)

**Comment by harrism** [comment]
Still stops early if there is a bracket in one of the quoted names: https://regex101.com/r/J3sjin/6

In any case, I would prefer to debug this (using regex101, for example) than a bunch of nested loops.

**Comment by OlivierNV** [comment]
I prefer the nested loops, but that's just me (at least it's clear what's actually being done, no need for 3rd party debugging tools). Long regex expressions look like the winner of an obfuscated C contest to me :)

**Comment by OlivierNV** [comment]
Replaced the regex with a simple custom implementation to also handle escaped quotes (takes demonstrably less time than figuring out if it would be reasonably doable with regex)

**Comment by harrism** [comment]
Here's a fairly simple regex that works: https://regex101.com/r/ui5iFi/9 , and doesn't rely on lookbehind.
```
"index_columns"\s*:\s*\[((?:(['"])(?:.(?!\2))*.\2,?\s*)*)\]
```

This works for single or double quotes. However if there are any escaped quotes (`\"`) in the strings it does not work -- adding that requires lookbehind support but I don't think C++ regex has that.

If that's not a requirement I think the above is quote clean.

**Comment by OlivierNV** [comment]
Also doesn't seem to work if there is a space between the various syntax elements like : or [] and the quote (maybe adding various " ?" might cure that ?), but I double-checked and we do need to replace \" with ", but we do not need single quotes (the pandas JSON syntax apparently requires double quotes)

**Comment by harrism** [comment]
Now supports white space after the open `[` (arbitrary whitespace was already supported everywhere else). And if only double quotes are required it's even simpler. https://regex101.com/r/ui5iFi/10

Hmmm, I guess `std::json_string` isn't a real thing. :)

**Comment by harrism** [comment]
OK, this one: https://regex101.com/r/ui5iFi/11
* Handles empty brackets after `"index_columns"`.
* Handles empty strings within the list.
* Handles escaped quotes, brackets, etc. within the quoted string.
* Does not use (negative) lookbehind.

```
  // Captures a list of quoted strings found inside square brackets after `"index_columns":`
  // Inside quotes supports newlines, brackets, escaped quotes, etc. 
  // One-liner regex:
  // "index_columns"\s*:\s*\[\s*((?:"(?:|(?:.*?(?![^\\]")).?)[^\\]?",?\s*)*)\]
  // Documented below.
  std::regex index_columns_expr{
    R"("index_columns"\s*:\s*\[\s*)"    // match preamble, opening square bracket, whitespace
      R"(()"                            // Open first capturing group
        R"((?:")"                       // Open non-capturing group match opening quote
          R"((?:|(?:.*?(?![^\\]")).?))" // match empty string or anything between quotes
          R"([^\\]?")"                  // Match closing non-escaped quote
          R"(,?\s*)"                    // Match optional comma and whitespace
        R"()*)"                         // Close non-capturing group and repeat 0 or more times
      R"())"                            // Close first capturing group
    R"(\])"                             // Match closing square brackets
  };
```

In general I'm strongly in favor of using C++ standard libraries over complex nested loops. Documenting the regex as above makes it understandable.

Test:
https://godbolt.org/z/Q2zPoZ

**Comment by OlivierNV** [comment]
Heh. Regex it is, then.

**Comment by OlivierNV** [comment]
@harrism  While the new regex does capture the escaped quotes as part of the pandas_index, this results in the column name being truncated at the first escape sign: for example, in your godbolt link, replacing the first column name by `"flo\"ats"` results in the following output:
```
strings
flo
doubles
```
instead of the expexted
```
strings
flo"ats
doubles
```

**Comment by OlivierNV** [comment]
@harrism You want to take another crack at it, or should I go ahead with the simple-but-raw loop ?

**Comment by harrism** [comment]
Note there are two regexes required in the C++. This is clearly just caused by not updating the simpler regex to parse the individual quoted strings. Just replace it with `R"(\"((?:\\.|[^\"])*)\")"`

https://godbolt.org/z/hBk4Gu

**Comment by OlivierNV** [comment]
That last find the correct word boundary, but it does not remove the `\` in `\"` (it needs to be removed)

**Comment by harrism** [comment]
https://godbolt.org/z/3VURrU

**Comment by OlivierNV** [comment]
Made the change back to regex. FWIW, the end result is to me a very good example of why ideological dogma and demands for conformity are not a good thing in software engineering.

**Comment by harrism** [comment]
I don't think the connection to #4718 is relevant. libcudf is large because it has to support many algorithms with combinations of types and operators. A large part of #4718 is temporary, since we have two copies of everything currently.

**Pull request merged**